src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs

```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
+                    arg.PrimitiveValue = new PrimitiveValue() { Byte4 = BitConverter.SingleToInt32Bits(parser.GetR4()) };
+                    break;
+                case CustomAttributeEncoding.Double:
+                    arg.PrimitiveValue = new PrimitiveValue() { Byte8 = BitConverter.DoubleToInt64Bits(parser.GetR8()) };
+                    break;
+                case CustomAttributeEncoding.String:
+                case CustomAttributeEncoding.Type:
+                    arg.StringValue = parser.GetString();
+                    break;
+                case CustomAttributeEncoding.Array:
+                {
+                    arg.ArrayValue = null;
+                    int len = parser.GetI4();
+                    if (len != unchecked((int)0xffffffff)) // indicates array is null - ECMA-335 II.23.3.
+                    {
+                        attributeType = new CustomAttributeType(
+                            attributeType.EncodedArrayType,
+                            CustomAttributeEncoding.Undefined, // Array type
+                            attributeType.EncodedEnumType,
+                            attributeType.EnumType);
+                        arg.ArrayValue = new CustomAttributeEncodedArgument[len];
+                        for (int i = 0; i < len; ++i)
+                        {
+                            arg.ArrayValue[i] = ParseCustomAttributeValue(ref parser, attributeType, module);
+                        }
+                    }
+                    break;
+                }
+                default:
+                    throw new BadImageFormatException();
+            }
+
+            return arg;
+        }
+
+        private static CustomAttributeType ParseCustomAttributeType(ref CustomAttributeDataParser parser, RuntimeModule module)
         {
-            m_type = type;
-            m_encodedArgument = default;
+            CustomAttributeEncoding arrayTag = CustomAttributeEncoding.Undefined;
+            CustomAttributeEncoding enumTag = CustomAttributeEncoding.Undefined;
+            Type? enumType = null;
+
+            CustomAttributeEncoding tag = parser.GetTag();
+            if (tag is CustomAttributeEncoding.Array)
+            {
+                arrayTag = parser.GetTag();
+            }
+
+            // Load the enum type if needed.
+            if (tag is CustomAttributeEncoding.Enum
+                || (tag is CustomAttributeEncoding.Array
+                    && arrayTag is CustomAttributeEncoding.Enum))
+            {
+                // We cannot determine the underlying type without loading the enum.
+                string? enumTypeMaybe = parser.GetString();
+                if (enumTypeMaybe is null)
+                {
+                    throw new BadImageFormatException();
+                }
+
+                enumType = TypeNameParser.GetTypeReferencedByCustomAttribute(enumTypeMaybe, module);
+                if (enumType is null || !enumType.IsEnum)
```

Comment:
```suggestion
                if (!enumType.IsEnum)
```
This cannot return null